### ISSUE TITLE
Arvo does not compile with %pure-mint

### ISSUE THREAD

**Issue by ponnys-podfer** [urbit/arvo#325]
Steps to reproduce:

1) Start a fakezod with the new `++velo` reboot sequence from master.

2) Comment out the hint on `++mint:ut`, along with something that probably doesn't matter, like the first error crash. (Other changes include adding another case to the `?-(gen ...)` case.

```
--- ../arvo/sys/hoon.hoon	2016-12-19 21:00:57.252257371 -0800
+++ home/sys/hoon.hoon	2016-12-26 17:17:01.419324902 -0800
@@ -7204,14 +7204,14 @@
     ?:(=(neg gen) sut $(gen neg))
   ::
   ++  mint
-    ~/  %mint
+    ::  ~/  %mint
     |=  {gol/span gen/twig}
     ^-  {p/span q/nock}
     ~&  %pure-mint
     |^  ^-  {p/span q/nock}
     ?:  ?&(=(%void sut) !?=({$dbug *} gen))
-      ?.  |(!vet ?=({$lost *} gen) ?=({$fail *} gen))
-        ~>(%mean.[%leaf "mint-vain"] !!)
+      ::  ?.  |(!vet ?=({$lost *} gen) ?=({$fail *} gen))
+      ::    ~>(%mean.[%leaf "mint-vain"] !!)
       [%void %0 0]
     ?-    gen
     ::
```
     
3) This is enough to cause a nest-fail during the automated |reset.

```
%hoon-compile ~polnem-fitped]
hoon-assembly
%pure-mint
[%compile-arvo ~tamdes-tiplup ~lagluc-podtyv]
%pure-mint
< ... lots of %pure-mint ... >
%pure-mint
nest-fail
/:<[365 7].[369 54]>
/:<[363 5].[384 7]>
/:<[361 5].[384 7]>
/:<[358 5].[384 7]>
/:<[357 5].[384 7]>
/:<[285 3].[393 5]>
/:<[284 3].[393 5]>
/:<[11 1].[394 3]>
/:<[10 1].[394 3]>
/:<[9 1].[649 3]>
/:<[5 1].[649 3]>
/:<[4 1].[649 3]>
/:<[588 11].[588 38]>
/:<[588 7].[588 38]>
/:<[588 3].[616 44]>
/:<[587 3].[616 44]>
/:<[583 3].[616 44]>
/:<[579 3].[616 44]>
/:<[571 3].[616 44]>
/:<[567 3].[616 44]>
/:<[566 3].[616 44]>
/:<[562 3].[616 44]>
/:<[558 3].[616 44]>
/:<[557 3].[616 44]>
/:<[553 3].[616 44]>
: /~zod/home/4/sys/hoon/hoon
```

(It would be nice if there were paths in the stack trace though.)

The jet version of mint compiles this fine.

**Comment by cgyarvin** [comment]
Interesting!  It would be neat to track down the discrepancy.  And probably
not *that* hard.  But...

The truth is that ++ut needs a rewrite, not because of this, but because
for obvious reasons it's the oldest code in the system -- and it relies
existentially on the partial-evaluation jet caching that assumes a
conservative worklist algorithm (in which the worklist isn't part of the
cache index).  I've concluded that this mechanism can't be made compatible
in a foolproof way, and in any case it's very very bad to *need* jet
propulsion for anything but simple arithmetic.  So ++ut needs an explicit
cache, not an implicit one dropped in at the jet level.  This is not a hard
problem, but I didn't know how to do it right when I wrote ++ut.

~rolsut-rospub has made some efforts in the direction of rewriting ++ut,
but it's a problem of more academic interest than helping-users interest.
On the other hand, whoever does it first will basically be the prime
minister of Hoon...
